cpp/src/shuffler/shuffler.cpp

```diff
@@ -457,27 +468,46 @@ void Shuffler::shutdown() {
     }
 }
 
+detail::Chunk Shuffler::create_chunk(
+    PartID pid, PackedData&& packed_data, std::shared_ptr<Buffer::Event> event
+) {
+    return detail::Chunk::from_packed_data(
+        get_new_cid(), pid, std::move(packed_data), std::move(event), stream_, br_
+    );
+}
+
 void Shuffler::insert_into_ready_postbox(detail::Chunk&& chunk) {
     auto& log = comm_->logger();
     log.trace("insert_into_outbox: ", chunk);
-    auto pid = chunk.pid;
-    if (chunk.expected_num_chunks) {
-        finish_counter_.move_goalpost(chunk.pid, chunk.expected_num_chunks);
+    // TODO: Now chunk could have multiple messages, so we need to unwrap them and
+    // insert them into the postbox individually because ready_postbox has partition
+    // id as key.
+    auto pid = chunk.part_id(0);
+    if (chunk.is_control_message(0)) {
+        finish_counter_.move_goalpost(pid, chunk.expected_num_chunks(0));
     } else {
-        ready_postbox_.insert(std::move(chunk));
+        if (chunk.n_messages() == 1) {
+            ready_postbox_.insert(std::move(chunk));
+        }
+        // else
+        // TODO: ready_postbox has partition id as key, so we need unwrap all messages
+        // in the chunk and insert them into the postbox.
     }
     finish_counter_.add_finished_chunk(pid);
 }
 
 void Shuffler::insert(detail::Chunk&& chunk) {
     {
         std::lock_guard const lock(outbound_chunk_counter_mutex_);
-        ++outbound_chunk_counter_[chunk.pid];
+        // TODO: There are multiple partitions in the chunk. So, do this for each
+        // partition.
+        ++outbound_chunk_counter_[chunk.part_id(0)];
     }
-    if (partition_owner(comm_, chunk.pid) == comm_->rank()) {
-        if (chunk.gpu_data) {
-            statistics_->add_bytes_stat("shuffle-payload-send", chunk.gpu_data->size);
-            statistics_->add_bytes_stat("shuffle-payload-recv", chunk.gpu_data->size);
+    // TODO: Guarantee that all messages in the chunk map to the same key (rank).
```

Comment:
As above, do this in the constructor of `Chunk` then you don't need to litter the rest of the code with checks.

Reply:
Are you suggesting that we add a mapping function to the Chunk ctr?